### PR TITLE
Fix typespec for put_new_layout/2 and put_new_view/2

### LINKED ISSUE #5031: phoenix 1.7 dialyzer error in put_new_layout
When generating a fresh phoenix project with the latest version, I get the dialyzer error shown below.

I believe the error is caused by the mismatch in the second argument:

`[{:html, {PhxdemoWeb.Layouts, :app}}]` vs current typespec `{atom(), binary() | atom()} | false`

```
lib/phxdemo_web/controllers/page_controller.ex:1:call
The function call will not succeed.

Phoenix.Controller.put_new_layout(
  _ :: %Plug.Conn{
    :adapter => {atom(), _},
    :assigns => %{atom() => _},
    :body_params => %Plug.Conn.Unfetched{:aspect => atom(), binary() => _},
    :cookies => %Plug.Conn.Unfetched{:aspect => atom(), binary() => _},
    :halted => boolean(),
    :host => binary(),
    :method => binary(),
    :owner => pid(),
    :params => %Plug.Conn.Unfetched{:aspect => atom(), binary() => _},
    :path_info => [binary()],
    :path_params => %{binary() => binary() | [any()] | map()},
    :port => char(),
    :private => %{atom() => _},
    :query_params => %Plug.Conn.Unfetched{
      :aspect => atom(),
      binary() => binary() | [any()] | map()
    },
    :query_string => binary(),
    :remote_ip =>
      {byte(), byte(), byte(), byte()}
      | {char(), char(), char(), char(), char(), char(), char(), char()},
    :req_cookies => %Plug.Conn.Unfetched{:aspect => atom(), binary() => binary()},
    :req_headers => [{_, _}],
    :request_path => binary(),
    :resp_body =>
      nil
      | binary()
      | maybe_improper_list(
          binary() | maybe_improper_list(any(), binary() | []) | byte(),
          binary() | []
        ),
    :resp_cookies => %{binary() => map()},
    :resp_headers => [{_, _}],
    :scheme => :http | :https,
    :script_name => [binary()],
    :secret_key_base => nil | binary(),
    :state => :chunked | :file | :sent | :set | :set_chunked | :set_file | :unset,
    :status => nil | non_neg_integer()
  },
  [{:html, {PhxdemoWeb.Layouts, :app}}]
)

breaks the contract
(Plug.Conn.t(), {atom(), binary() | atom()} | false) :: Plug.Conn.t()

________________________________________________________________________________
lib/phxdemo_web/controllers/page_controller.ex:2:no_return
Function call/2 has no local return.
```

### PR DESCRIPTION
**Pull request by dvic**
Fixes #5031

### CODE REVIEW

**Comment by dvic** [comment]
Not 100% sure if this is correct, I tried to deduce from the source code what the correct typespec should be.

File: lib/phoenix/controller.ex [hunk 1]
```diff
@@ -474,12 +474,14 @@ defmodule Phoenix.Controller do
     end)
   end
 
+  @type view :: {format :: atom, view :: atom}
```
Comment (josevalim) [inline lib/phoenix/controller.ex:477]: This makes the type deeply recursive, which is not what we want. The format is also only supported as a keyword list.

File: lib/phoenix/controller.ex [hunk 2]
```diff
@@ -474,12 +474,14 @@ defmodule Phoenix.Controller do
...
+
```
Comment (josevalim) [inline lib/phoenix/controller.ex:478]:
```suggestion
```

**Comment by dvic** [comment]
The current version gives the error below. I believe the typespec is missing `false` as an option?

```
lib/demo_web/controllers/page_controller.ex:1:call
The function call will not succeed.

Phoenix.Controller.put_layout(
  _ :: %Plug.Conn{
    :adapter => {atom(), _},
    :assigns => %{atom() => _},
    :body_params => %Plug.Conn.Unfetched{:aspect => atom(), binary() => _},
    :cookies => %Plug.Conn.Unfetched{:aspect => atom(), binary() => _},
    :halted => false,
    :host => binary(),
    :method => binary(),
    :owner => pid(),
    :params => %Plug.Conn.Unfetched{:aspect => atom(), binary() => _},
    :path_info => [binary()],
    :path_params => %{
      binary() =>
        binary()
        | [binary() | [any()] | %{binary() => _}]
        | %{binary() => binary() | [any()] | %{binary() => _}}
    },
    :port => char(),
    :private => %{atom() => _},
    :query_params => %Plug.Conn.Unfetched{
      :aspect => atom(),
      binary() =>
        binary()
        | [binary() | [any()] | %{binary() => _}]
        | %{binary() => binary() | [any()] | %{binary() => _}}
    },
    :query_string => binary(),
    :remote_ip =>
      {byte(), byte(), byte(), byte()}
      | {char(), char(), char(), char(), char(), char(), char(), char()},
    :req_cookies => %Plug.Conn.Unfetched{:aspect => atom(), binary() => binary()},
    :req_headers => [{binary(), binary()}],
    :request_path => binary(),
    :resp_body =>
      nil
      | binary()
      | maybe_improper_list(
          binary() | maybe_improper_list(any(), binary() | []) | byte(),
          binary() | []
        ),
    :resp_cookies => %{binary() => map()},
    :resp_headers => [{binary(), binary()}],
    :scheme => :http | :https,
    :script_name => [binary()],
    :secret_key_base => nil | binary(),
    :state => :chunked | :file | :sent | :set | :set_chunked | :set_file | :unset,
    :status => nil | non_neg_integer()
  },
  false
)

breaks the contract
(Plug.Conn.t(), [{format :: atom(), layout()}]) :: Plug.Conn.t()
```

**Comment by josevalim** [comment]
Correct! As mentioned above, you must use keyword lists with the format when managing the layout. :)

**Comment by josevalim** [comment]
Actually, I will change that around. I will merge and address it soon.

**Comment by dvic** [comment]
Ah wait, I did not realise there was this line in the generated code: `plug :put_layout, false when action in [:home]`. So that should become `plug :put_layout, html: false when action in [:home]`?

**Comment by josevalim** [comment]
I am almost sure that line was fully removed in recent generators, correct?

**Comment by dvic** [comment]
> I am almost sure that line was fully removed in recent generators, correct?

I just checked by generating a new project with the latest version: and indeed, the line is gone.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**